libcudacxx/include/cuda/__iterator/transform_iterator.h

```diff
@@ -0,0 +1,501 @@
+//!
+//! @code
+//! #include <cuda/iterator>
+//! #include <thrust/device_vector.h>
+//!
+//! struct square_root
+//! {
+//!   __host__ __device__
+//!   float operator()(float x) const
+//!   {
+//!     return sqrtf(x);
+//!   }
+//! };
+//!
+//! int main()
+//! {
+//!   thrust::device_vector<float> v(4);
+//!   v[0] = 1.0f;
+//!   v[1] = 4.0f;
+//!   v[2] = 9.0f;
+//!   v[3] = 16.0f;
+//!
+//!   using FloatIterator = thrust::device_vector<float>::iterator;
+//!
+//!   cuda::transform_iterator iter(v.begin(), square_root{});
+//!
+//!   *iter;   // returns 1.0f
+//!   iter[0]; // returns 1.0f;
+//!   iter[1]; // returns 2.0f;
+//!   iter[2]; // returns 3.0f;
+//!   iter[3]; // returns 4.0f;
+//!
+//!   // iter[4] is an out-of-bounds error
+//! }
+//! @endcode
+//!
+//! This next example demonstrates how to use a \p transform_iterator with the \p thrust::reduce function to compute the
+//! sum of squares of a sequence. We will create temporary \p transform_iterators with the \p make_transform_iterator
+//! function in order to avoid explicitly specifying their type:
+//!
+//! @code
+//! #include <cuda/iterator>
+//! #include <thrust/device_vector.h>
+//! #include <thrust/reduce.h>
+//! #include <iostream>
+//!
+//! struct square
+//! {
+//!   __host__ __device__
+//!   float operator()(float x) const
+//!   {
+//!     return x * x;
+//!   }
+//! };
+//!
+//! int main()
+//! {
+//!   // initialize a device array
+//!   thrust::device_vector<float> v(4);
+//!   v[0] = 1.0f;
+//!   v[1] = 2.0f;
+//!   v[2] = 3.0f;
+//!   v[3] = 4.0f;
+//!
+//!   float sum_of_squares =
+//!    thrust::reduce(cuda::make_transform_iterator(v.begin(), square{}),
+//!                   cuda::make_transform_iterator(v.end(),   square{}));
+//!
+//!   std::cout << "sum of squares: " << sum_of_squares << std::endl;
+//!   return 0;
+//! }
+//! @endcode
+//!
+//! The following example illustrates how to use the third template argument to explicitly specify the return type of
+//! the function.
+//!
+//! @code
+//! #include <cuda/iterator>
+//! #include <thrust/device_vector.h>
+//!
+//! struct square_root
+//! {
+//!   __host__ __device__
+//!   float operator()(float x) const
+//!   {
+//!     return sqrtf(x);
+//!   }
+//! };
+//!
+//! int main()
+//! {
+//!   thrust::device_vector<float> v(4);
+//!   v[0] = 1.0f;
+//!   v[1] = 4.0f;
+//!   v[2] = 9.0f;
+//!   v[3] = 16.0f;
+//!
+//!   using FloatIterator = thrust::device_vector<float>::iterator;
+//!
+//!   // note: float result_type is specified explicitly
+//!   cuda::transform_iterator iter(v.begin(), square_root{});
+//!
+//!   *iter;   // returns 1.0f
+//!   iter[0]; // returns 1.0f;
+//!   iter[1]; // returns 2.0f;
+//!   iter[2]; // returns 3.0f;
+//!   iter[3]; // returns 4.0f;
+//!
+//!   // iter[4] is an out-of-bounds error
+//! }
+//! @endcode
+template <class _Iter, class _Fn>
+class transform_iterator : public __transform_iterator_category_base<_Iter, _Fn>
```

Comment:
Note: Thrust has the reverse order of template parameters. This version here is constent with ranges.

Reply:
This has been a thorn in my side, because I needed to change the constructors to be consistent with the thrust iterators

Reply:
Leave it. I like your version better.